docs/designdoc/allow_customize_containers.md

```diff
@@ -0,0 +1,104 @@
+# Allow users to customize containers
+
+## Context
+
+Currently, MOCO has containers that are automatically added by the system in addition to containers added by the user.
+(e.g. `agent`, `moco-init` etc...)
+Containers automatically added by the system do not allow user customization.
+
+There are cases where users want to customize container resources, as in Issue [cybozu-go/moco#235](https://github.com/cybozu-go/moco/issues/235).
+This design document examines how users can customize containers that are automatically added by the system.
+
+## Goals
+
+* No breaking changes
+* Allows users to customize containers that are automatically added by the system
+* Extendable functionality for future customization beyond container resources
+* Users can only customize authorized fields
+  * Do not allow customization of fields that would make operation untenable (e.g. `command`)
+
+## Non-goals
+
+* No customization provided at the time of implementation other than requested container resources
+* No enhancement will be added to the `moco.cybozu.com/v1beta1` API.
+
+## ActualDesign
+
+Add `.spec.podTemplate.overwriteContainers` field to MySQLCluster.
+
+```yaml
+spec:
+  podTemplate:
+    spec:
+      containers:
+      - name: mysqld
+        image: quay.io/cybozu/mysql:8.0.28
+    overwriteContainers:
+    - name: agent
+      resources:
+        requests:
+          cpu: 50m
+```
+
+Pseudo-code:
+
+```go
+type PodTemplateSpec struct {
+	ObjectMeta
+	Spec                PodSpecApplyConfiguration
+	OverwriteContainers []OverwriteContainer // added
+}
+
+type OverwriteContainer struct {
+	Name      string // required
+	Resources *corev1ac.ResourceRequirementsApplyConfiguration
+}
+```
+
+`overwriteContainers` is a container definition with only name and customizable fields.
+moco-controller refers to `overwriteContainers` when creating containers and overwrites fields if the container name matches.
+
+`overwriteContainer` does not distinguish between initContainer and container, and the names of required fields are validated with the validation webhook.
+No merge logic is provided for container resources to avoid implicit value setting situations.
+
+Pros:
+
+* Expandable
+* Users can learn about customizable fields from MySQLCluster specs
+
+Cons:
+
+* If you want different customizable fields for different containers, you cannot express
+* Users need to know the list of containers to be added by the system
+  * Follow it in the documentation
+
+## AlternativesConsidered
```

Comment:
I prefer this alternative design.  It is at least explicit.
We can reduce potential user stress by setting the default resource limits.

Reply:
@ymmt2005 
I discussed this design with @zoetrope. And we prefer the first one.
We think it is easy to approach because the structure is similar to the `corev1.Container`.

Could we choose your proposed alternative design, which limits the `name` filed with the `Enum` marker?

https://book.kubebuilder.io/reference/markers/crd-validation.html
```
// +kubebuilder:validation:Enum=<[]any>
    specifies that this (scalar) field is restricted to the *exact* values specified here.
```

Reply:
@masa213f Not bad :-)

Reply:
@ymmt2005 
Thanks for checking!

@d-kuro 
Could you update the first design to limit with the enum marker?